feat(terminal): jobstart() term:// buffer name

justinmk opened the pull request:
## Problem:

The `term://` buffer-name generated by `jobstart(term=true)` is mostly broken, because it does not set the full command, which:
- is confusing
- will fail to "restore" (re-invoke) the terminal by visiting the `term://` buffer-name.

## Solution:
- Append the full cmd+args to the URI as a JSON array.
- The reverse case of `:edit term://...` still works as normal, since in that case the user is providing the argument delimiting (quoting).
- Improve the default `BufReadCmd term://*` handler:
  - Handles the JSON command form.
  - Rewrite as Lua (missing `|endif` since c5b34fa55483d84d1de32937ffff0b7cf1aeba78 🙈)
- Support URI `#fragment` https://github.com/neovim/neovim/issues/3278
    - Alternative: `term+1234://` ? [ref](https://github.com/neovim/neovim/pull/36564)

Fix https://github.com/neovim/neovim/issues/3278
Fix https://github.com/neovim/neovim/issues/14525
Fix https://github.com/neovim/neovim/issues/32932
Ref https://github.com/neovim/neovim/issues/31494

Before:

```
:lua vim.fn.jobstart({'gh','status'},{term=true})
  => term://~//67329:/usr/local/bin/gh             # broken

:edit term://gh status
  => term://~//67366:gh status
```

After:

```
:lua vim.fn.jobstart({'gh','status'},{term=true})
  => term://~//67314:["/usr/local/bin/gh", "status"]

:edit term://gh status
  => term://~//67366:gh status                     # no change
```

## Todo

- Support URI `#fragment`
- ✅ `:edit term://...` should preserve its name?
- (unrelated to this PR) fix [leak](https://github.com/neovim/neovim/pull/35951#discussion_r2389709575) when `jobstart` reuses an existing terminal buffer.

justinmk committed:
fix(terminal): `jobstart(term=true)` term:// buf name

Problem:
The `term://` buffer-name generated by `jobstart(term=true)` is mostly
broken, because it does not set the full command, which:
- is confusing
- will fail to "restore" (re-invoke) the terminal by visiting the
  `term://` buffer-name.

Solution:
- Append the full cmd+args to the URI as a JSON array.
  - Future: URI v2: github.com/neovim/neovim/issues/3278
- Improve the default `BufReadCmd term://*` handler:
  - Handles the JSON command form.
  - Rewrite as Lua (missing `|endif` since c5b34fa55483d84d1de32937ffff0b7cf1aeba78 🙈)

Before:

    term://~/foo//80047:/path/to/shell-test

After:

    term://~/foo//80047:["/path/to/shell-test", "arg1", "arg2"]

justinmk commented on test/functional/terminal/edit_spec.lua:
```diff
@@ -61,4 +64,46 @@ describe(':edit term://*', function()
+    )
+
+    fn.jobwait({ jobid })
+    command('set nomodified')
+    -- Try a command with complex quoted args.
+    fn.jobstart({ testprg('shell-test'), 'arg1', [[["arg 2"] "with" 'quotes']] }, { term = true })
```
Forcing 'nomodified' exposes a leak. This is not technically "supported" but maybe we should.

```
= Indirect leak of 80000 byte(s) in 1 object(s) allocated from:
=     #0 0x555f2a37cda3 in malloc (/home/runner/work/neovim/neovim/build/bin/nvim+0xd2dda3) (BuildId: 0223c81575fe8615e437fe31bdd2262c6d6305c6)
=     #1 0x555f2b01d774 in try_malloc /home/runner/work/neovim/neovim/src/nvim/memory.c:113:15
=     #2 0x555f2b01d9d4 in xmalloc /home/runner/work/neovim/neovim/src/nvim/memory.c:147:15
=     #3 0x555f2b73bf5e in terminal_open /home/runner/work/neovim/neovim/src/nvim/terminal.c:532:23
=     #4 0x555f2a675a92 in channel_terminal_open /home/runner/work/neovim/neovim/src/nvim/channel.c:832:3
=     #5 0x555f2a956664 in f_jobstart /home/runner/work/neovim/neovim/src/nvim/eval/funcs.c:4067:5
```

justinmk committed:
feat: don't expand URIs passed to :argadd, :badd

oni-link commented:
> ## Problem:
> 
> The `term://` buffer-name generated by `jobstart(term=true)` is mostly broken, because it does not set the full command, which:
> 
>     * is confusing
> 
>     * will fail to "restore" (re-invoke) the terminal by visiting the `term://` buffer-name.
> 

Is this relevant here: What if the full command contains a `rm ...` and we "restore" (re-invoke) the terminal now from the full command? Suprise deletion or expected behavior?

justinmk commented:
> What if the full command contains a `rm ...` and we "restore" (re-invoke) the terminal now from the full command?

Two things to note:

- if the last arg is `...` then the command is not run (there is a test for this)
- but your question in general is valid. it also applies to the existing behavior, which just is kind of broken but still could run an (incomplete) destructive operation

So either we should prompt before actually "restoring" (re-invoking) the terminal, or just don't attempt to re-invoke it.

justinmk committed:
fix(session): avoid `:buffer [name]`, it fails with `term://…/["x"]` names

Problem:

    ERROR test/functional/ex_cmds/mksession_spec.lua @ 36: :mksession restores same :terminal buf in splits
    test\functional\testnvim.lua:141: D:\a\neovim\neovim\Xtest-functional-ex_cmds-mksession_spec.vim,
    line 39: Vim(buffer):E94: No matching buffer for term://D:/a/neovim/neovim//3788:["C://Windows//system32//cmd.exe"]

Solution:
Use `:buffer [num]` instead.
This also reverts the `:file` hack added in 1e103b3c12597a9dd2f20d45686822ab6ee089b0

justinmk committed:
feat: never expand() URIs

justinmk committed:
WIP

Willaaaaaaa commented on src/nvim/eval/funcs.c:
```diff
@@ -3687,8 +3690,21 @@ void f_jobstart(typval_T *argvars, typval_T *rettv, EvalFuncData fptr)
+      // Serialize command as a JSON array: ["cmd", "arg1", "arg 2", …].
+      // URI v2: https://github.com/neovim/neovim/issues/3278
+      cmdjson = shell_argv_to_json(argv, true);
+      cmd = shell_argv_to_json(argv, true);
```
This line causes memory leak and makes no sense. It should be deleted.

justinmk replied:
oh, this might be a merge mistake, thanks

Willaaaaaaa commented on runtime/lua/vim/_core/defaults.lua:
```diff
@@ -573,9 +573,42 @@ do
+        return
+      end
+      -- NOTE: the `(\d+:)` is ignored, for back-compat (old PID placement).
+      -- PID is now expected in the `#fragment` at the end.
```
I love this idea! So it's like `term://{spawn_time_cwd}//{cmd}#{pid}[#{fragment}]`? Seems some existing implementations and docs are conflict with this wonderful idea.
